src/cephadm/cephadm.py

```diff
@@ -6129,49 +6172,114 @@ def get_deployment_container(ctx: CephadmContext,
 
 
 @default_image
+@deprecated_command
 def command_deploy(ctx):
     # type: (CephadmContext) -> None
-    daemon_type, daemon_id = ctx.name.split('.', 1)
+    _common_deploy(ctx)
+
+
+def read_configuration_source(ctx: CephadmContext) -> Dict[str, Any]:
+    """Read a JSON configuration based on the `ctx.source` value."""
+    source = '-'
+    if 'source' in ctx and ctx.source:
+        source = ctx.source
+    if source == '-':
+        config_data = json.load(sys.stdin)
+    else:
+        with open(source, 'rb') as fh:
+            config_data = json.load(fh)
+    logger.debug('Loaded deploy configuration: %r', config_data)
+    return config_data
 
-    lock = FileLock(ctx, ctx.fsid)
-    lock.acquire()
 
+def apply_deploy_config_to_ctx(
+    config_data: Dict[str, Any],
+    ctx: CephadmContext,
+) -> None:
+    """Bind properties taken from the config_data dictionary to our ctx,
+    similar to how cli options on `deploy` are bound to the context.
+    """
+    ctx.name = config_data['name']
+    if 'image' in config_data:
+        ctx.image = config_data['image']
+    if 'fsid' in config_data:
+        ctx.fsid = config_data['fsid']
+    if 'meta' in config_data:
+        ctx.meta_properties = config_data['meta']
+    if 'config_blobs' in config_data:
+        ctx.config_blobs = config_data['config_blobs']
+        ctx.config_json = config_data['config_blobs']
+    # this is quite the hack!  this is a workaround for the fact that
+    # cephadm orch module knows about "args" not what they represent
+    # and so this lets us (try to) smooth the transition. Ideally, these
+    # would be set using key-value pairs in 'params' below.
+    if 'deploy_arguments' in config_data:
+        parser = argparse.ArgumentParser(add_help=False, exit_on_error=False)
```

Comment:
@phlogistonjohn did an initial test run with just this PR in it 

https://pulpito.ceph.com/adking-2023-06-01_03:43:37-orch:cephadm-wip-adk4-testing-2023-05-31-2041-distro-default-smithi/

88 failures, didn't look through all of them, but the few I did look at I saw

```
Traceback (most recent call last):
File "/usr/lib64/python3.6/runpy.py", line 193, in _run_module_as_main
  "__main__", mod_spec)
File "/usr/lib64/python3.6/runpy.py", line 85, in _run_code
  exec(code, run_globals)
File "/tmp/tmp1xjy1b_9.cephadm.build/__main__.py", line 10229, in <module>
File "/tmp/tmp1xjy1b_9.cephadm.build/__main__.py", line 10217, in main
File "/tmp/tmp1xjy1b_9.cephadm.build/__main__.py", line 6236, in command_deploy_from
File "/tmp/tmp1xjy1b_9.cephadm.build/__main__.py", line 6219, in apply_deploy_config_to_ctx
TypeError: __init__() got an unexpected keyword argument 'exit_on_error'
```

which seems to be coming from this line

Reply:
I updated the code but, while working on it, I started wondering how we hit this because it should have only called `argparse.ArgumentParser` *IF*  deploy_arguments  were present and the mgr module shouldn't have been even setting values there any more. Regardless, even the intermediate code no longer tries passing  exit_on_error.